polyglot/piranha/src/piranha.rs

```diff
@@ -59,76 +72,146 @@ impl SourceCodeUnit {
-        let next_rules_by_scope = rules_store.get_next(&current_rule, self.substitutions());
-
-        // Adds "Method" and "Class" rules to the stack
-        self.add_rules_to_stack(
-          &next_rules_by_scope,
-          current_edit,
-          rules_store,
-          &mut next_rules_stack,
-        );
+    let scope_node = self.get_scope_node(scope_query, rule_store);
 
-        // Add Global rules as seed rules
-        for r in &next_rules_by_scope[GLOBAL] {
-          rules_store.add_to_global_rules(r, self.substitutions());
-        }
+    let mut query_again = false;
 
-        // Process the parent
-        // Find the rules to be applied in the "Parent" scope that match any parent (context) of the changed node in the previous edit
-        if let Some(edit) = Rule::get_edit_for_context(
-          &self.clone(),
-          current_edit.start_byte,
-          current_edit.new_end_byte,
-          rules_store,
-          &next_rules_by_scope[PARENT],
-        ) {
-          #[rustfmt::skip]
-          info!( "{}", format!( "Cleaning up the context, by applying the rule - {}", edit.matched_rule().name()).green());
-          // Apply the matched rule to the parent
-          current_edit = self.apply_edit(&edit, parser);
-          current_rule = edit.matched_rule();
-          // Add the (tag, code_snippet) mapping to substitution table.
-          self.add_to_substitutions(edit.matches());
-        } else {
-          // No more parents found for cleanup
-          break;
-        }
+    // When rule is a "rewrite" rule :
+    // Update the first match of the rewrite rule
+    // Add mappings to the substitution
+    // Propagate each applied edit. The next rule will be applied relative to the application of this edit. 
+    if !rule.is_match_only_rule() {
+      if let Some(edit) = rule.get_edit(&self.clone(), rule_store, scope_node, true) {
+        self.rewrites_mut().push(edit.clone());
+        query_again = true;
+
+        // Add all the (code_snippet, tag) mapping to the substitution table.
+        self.add_to_substitutions(edit.matches());
+
+        // Apply edit_1
+        let applied_ts_edit = self.apply_edit(&edit, parser);
+
+        self.propagate(applied_ts_edit, rule, rule_store, parser);
+      }
+    }
+    // When rule is a "match-only" rule :
+    // Get all the matches
+    // Add mappings to the substitution
+    // Propagate each match. Note that,  we pass a identity edit (where old range == new range) in to the propagate logic. 
+    // The next edit will be applied relative to the identity edit. 
+    else {
+      for m in rule.get_matches(&self.clone(), rule_store, scope_node, true) {
+        self.matches_mut().push((rule.clone(), m.clone()));
+
+        self.add_to_substitutions(m.matches());
+        
+        // Create identity edit, that replaces a range with itself. 
+        let identity_edit = InputEdit {
```

Comment:
Fine to do in a follow up PR, but remember offline discussion about naming this something other than edit, specially if not actually applied to the source.

Reply:
Also, I don't fully see where the check is that distinguishes between a real edit and this before applying. Can I get a quick rundown of the logic here?

Reply:
One line [105](https://github.com/uber/piranha/pull/212/files/9cf8a1f970527f37d341e9f4c5fae0d1ac260266#diff-31027e2b569bb16bf28a5b188a0ca6edda8c55cda1ff87475f47e0366a84d778R105) we check if the rule is "rewrite rule"  - `!rule.is_match_only_rule()`. So if it is a rewrite rule : 
(i) It will get the `Edit`
(ii) Update the substitution table 
(iii)  apply it to the source code unit ([l. 114]
(iv) Propagate the edit 
(https://github.com/uber/piranha/pull/212/files/9cf8a1f970527f37d341e9f4c5fae0d1ac260266#diff-31027e2b569bb16bf28a5b188a0ca6edda8c55cda1ff87475f47e0366a84d778R114) ) - `apply_edit`.

IF its a match-only rule, we get all the matches and for each match : 
(i) Update substitution table 
(ii) Create this identity edit 
(iii) Propagate the identity edit

Reply:
I have done the refactoring to make the intention more clear. The change was simple and straightforward.,